build/builder.go

```diff
@@ -942,6 +945,40 @@ func sortDocuments(todo []*zoekt.Document) {
 	}
 }
 
+// sortDocuments2 sorts zoekt.Document according to their Scores. In general,
+// documents can have a nil score vector if the document to be indexed was
+// introduced after the ranking took place. A nil score vector translates to the
+// lowest possible rank.
+func sortDocuments2(rs []*zoekt.Document) {
+	sort.Slice(rs, func(i, j int) bool {
+		r1 := rs[i].Scores
+		r2 := rs[j].Scores
+
+		if r1 == nil && r2 == nil {
+			return false
+		}
+
+		if r1 == nil {
+			return false
+		}
+
+		if r2 == nil {
+			return true
+		}
+
+		for i := range r1 {
+			if r1[i] < r2[i] {
+				return true
```

Comment:
do we not want to in general inverse this condition? we want in general r1 < r2 as a vector to mean it is less important => when sorting we want the most important in the front. If I understand this implementation the most important is at the back.

Reply:
This is the same logic we have in `sortDocuments` and is the consequence of the `rank` func in builder.go. Eric has adopted the score vector semantics in the ranking service for now. Hopefully we can move from score vectors to ranks soon.

Reply:
sortDocuments and sortDocuments2 have exactly the same sorting logic. The only difference is that sortDocuments2 allows for empty/nil scores and (now) also score vectors of different lengths.